File: libraries/typescript/src/dashboard/store.ts
/**
 * In-memory metrics store for the local dashboard.
 *
 * Keeps the last `maxCalls` completed calls and tracks active calls.
 * Supports SSE event subscribers for real-time updates.
 *
 * Optional disk hydration: when `CallLogger` writes per-call records under
 * `<root>/calls/YYYY/MM/DD/<call_id>/metadata.json`, calling
 * `hydrate(logRoot)` on a fresh store rebuilds the in-memory list from those
 * files so the dashboard survives process restarts (the persistence is in
 * the JSONL/JSON files, the store is just a cache on top).
 */

import { EventEmitter } from 'events';
import * as fs from 'node:fs';
import * as path from 'node:path';
import { getLogger } from '../logger';

/** Snapshot of a call as held by the dashboard store. */
export interface CallRecord {
  call_id: string;
  caller: string;
  callee: string;
  direction: string;
  started_at: number;
  ended_at?: number;
  /**
   * Current lifecycle state: ``initiated`` (pre-registered), ``ringing``,
   * ``in-progress``, ``completed``, ``no-answer``, ``busy``, ``failed``,
   * ``canceled``, or ``webhook_error``.
   */
  status?: string;
  transcript?: Array<{ role: string; text: string; timestamp: number }>;
  turns?: unknown[];
  metrics?: Record<string, unknown> | null;
  [key: string]: unknown;
}

/** Server-Sent-Event payload broadcast by `MetricsStore` for live UI updates. */
export interface SSEEvent {
  type: string;
  data: Record<string, unknown>;
}

/** In-memory bounded ring buffer of recent calls plus active-call tracking. */
export class MetricsStore extends EventEmitter {
  private readonly maxCalls: number;
  private calls: CallRecord[] = [];
  private activeCalls: Map<string, CallRecord> = new Map();

  /**
   * Accepts either a numeric ``maxCalls`` (legacy positional — matches the
   * original TS API) or an options object ``{ maxCalls }`` to align with the
   * Python SDK's keyword-argument style. Plain literals also work:
   * ``new MetricsStore()`` / ``new MetricsStore(100)`` / ``new MetricsStore({ maxCalls: 100 })``.
   */
  constructor(maxCallsOrOpts: number | { maxCalls?: number } = 500) {
    super();
    this.maxCalls =
      typeof maxCallsOrOpts === 'number'
        ? maxCallsOrOpts
        : maxCallsOrOpts.maxCalls ?? 500;
  }

  private publish(eventType: string, data: Record<string, unknown>): void {
    this.emit('sse', { type: eventType, data } as SSEEvent);
  }

  /** Mark a call as in-progress (creates the row if it does not yet exist). */
  recordCallStart(data: Record<string, unknown>): void {
    const callId = (data.call_id as string) || '';
    if (!callId) return;

    const existing = this.activeCalls.get(callId);
    if (existing) {
      // Upgrade a pre-registered (``initiated``) outbound call to in-progress
      // without losing its from/to metadata. See BUG #06.
      existing.caller = (data.caller as string) || existing.caller;
      existing.callee = (data.callee as string) || existing.callee;
      existing.direction = (data.direction as string) || existing.direction;
      existing.status = 'in-progress';
      existing.turns = existing.turns || [];
    } else {
      const record: CallRecord = {
        call_id: callId,
        caller: (data.caller as string) || '',
        callee: (data.callee as string) || '',
        direction: (data.direction as string) || 'inbound',
        started_at: Date.now() / 1000,
        status: 'in-progress',
        turns: [],
      };
      this.activeCalls.set(callId, record);
    }

    this.publish('call_start', {
      call_id: callId,
      caller: (data.caller as string) || '',
      callee: (data.callee as string) || '',
      direction: (data.direction as string) || 'inbound',
    });
  }

  /**
   * Pre-register an outbound call before any webhook fires. Lets the
   * dashboard surface attempts that never reach media (no-answer, busy,
   * carrier-rejected). Mirrors the Python ``record_call_initiated``.
   */
  recordCallInitiated(data: Record<string, unknown>): void {
    const callId = (data.call_id as string) || '';
    if (!callId) return;
    if (this.activeCalls.has(callId)) return; // first writer wins

    const record: CallRecord = {
      call_id: callId,
      caller: (data.caller as string) || '',
      callee: (data.callee as string) || '',
      direction: (data.direction as string) || 'outbound',
      started_at: Date.now() / 1000,
      status: 'initiated',
      turns: [],
    };
    this.activeCalls.set(callId, record);
    this.publish('call_initiated', {
      call_id: callId,
      caller: record.caller,
      callee: record.callee,
      direction: record.direction,
      status: record.status,
    });
  }

  /**
   * Update the status of an active or completed call. Terminal states
   * (completed, no-answer, busy, failed, canceled, webhook_error) move the
   * row from active to completed so the UI freezes the live duration timer.
   */
  updateCallStatus(callId: string, status: string, extra: Record<string, unknown> = {}): void {
    if (!callId || !status) return;
    const TERMINAL = new Set(['completed', 'no-answer', 'busy', 'failed', 'canceled', 'webhook_error']);
    const active = this.activeCalls.get(callId);
    if (active) {
      active.status = status;
      Object.assign(active, extra);
      if (TERMINAL.has(status)) {
        const entry: CallRecord = {
          call_id: callId,
          caller: active.caller || '',
          callee: active.callee || '',
          direction: active.direction || 'outbound',
          started_at: active.started_at || 0,
          ended_at: Date.now() / 1000,
          status,
          metrics: null,
          ...extra,
        };
        this.activeCalls.delete(callId);
        this.calls.push(entry);
        if (this.calls.length > this.maxCalls) {
          this.calls = this.calls.slice(-this.maxCalls);
        }
      }
    } else {
      for (let i = this.calls.length - 1; i >= 0; i--) {
        if (this.calls[i].call_id === callId) {
          this.calls[i].status = status;
          Object.assign(this.calls[i], extra);
          break;
        }
      }
    }
    this.publish('call_status', { call_id: callId, status, ...extra });
  }

  /** Append a single conversation turn to an active call and broadcast it via SSE. */
  recordTurn(data: Record<string, unknown>): void {
    const callId = (data.call_id as string) || '';
    const turn = data.turn;
    if (!callId || turn == null) return;

    const active = this.activeCalls.get(callId);
    if (active) {
      if (!active.turns) active.turns = [];
      active.turns.push(turn);
    }

    this.publish('turn_complete', { call_id: callId, turn: turn as Record<string, unknown> });
  }

  /** Move a call from active to completed and persist its final metrics. */
  recordCallEnd(data: Record<string, unknown>, metrics?: Record<string, unknown> | null): void {
    const callId = (data.call_id as string) || '';
    if (!callId) return;

    const active = this.activeCalls.get(callId);
    this.activeCalls.delete(callId);

    // Preserve explicit status set by a statusCallback during the call
    // (e.g. "no-answer" from Twilio) — fall back to "completed" when the
    // row was still showing the normal "in-progress" state at hang-up.
    const activeStatus = active?.status;
    const resolvedStatus =
      activeStatus && activeStatus !== 'in-progress' ? activeStatus : 'completed';
    const entry: CallRecord = {
      call_id: callId,
      caller: (data.caller as string) || active?.caller || '',
      callee: (data.callee as string) || active?.callee || '',
      direction: active?.direction || (data.direction as string) || 'inbound',
      started_at: active?.started_at || 0,
      ended_at: Date.now() / 1000,
      transcript: (data.transcript as CallRecord['transcript']) || [],
      status: resolvedStatus,
      metrics: metrics ?? null,
    };

    this.calls.push(entry);
    if (this.calls.length > this.maxCalls) {
      this.calls = this.calls.slice(-this.maxCalls);
    }

    this.publish('call_end', {
      call_id: callId,
      metrics: entry.metrics ?? null,
    });
  }

  /** Return a window of completed calls in newest-first order. */
  getCalls(limit = 50, offset = 0): CallRecord[] {
    const ordered = [...this.calls].reverse();
    return ordered.slice(offset, offset + limit);
  }

  /** Look up a completed call by id (newest match wins). */
  getCall(callId: string): CallRecord | null {
    for (let i = this.calls.length - 1; i >= 0; i--) {
      if (this.calls[i].call_id === callId) return this.calls[i];
    }
    return null;
  }

  /** Look up an active call by id (returns undefined if not active or unknown). */
  getActive(callId: string): CallRecord | undefined {
    return this.activeCalls.get(callId);
  }

  /** Return all currently active (not yet ended) calls. */
  getActiveCalls(): CallRecord[] {
    return Array.from(this.activeCalls.values());
  }

  /** Compute summary statistics across the buffered call history. */
  getAggregates(): Record<string, unknown> {
    const totalCalls = this.calls.length;
    if (totalCalls === 0) {
      return {
        total_calls: 0,
        total_cost: 0,
        avg_duration: 0,
        avg_latency_ms: 0,
        cost_breakdown: { stt: 0, tts: 0, llm: 0, telephony: 0 },
        active_calls: this.activeCalls.size,
      };
    }

    let totalCost = 0;
    let totalDuration = 0;
    let totalLatency = 0;
    let latencyCount = 0;
    let costStt = 0;
    let costTts = 0;
    let costLlm = 0;
    let costTel = 0;

    for (const call of this.calls) {
      const m = call.metrics as Record<string, unknown> | null;
      if (!m) continue;
      const cost = (m.cost as Record<string, number>) || {};
      totalCost += cost.total || 0;
      costStt += cost.stt || 0;
      costTts += cost.tts || 0;
      costLlm += cost.llm || 0;
      costTel += cost.telephony || 0;
      totalDuration += (m.duration_seconds as number) || 0;
      const avgLat = (m.latency_avg as Record<string, number>) || {};
      const tMs = avgLat.total_ms || 0;
      if (tMs > 0) {
        totalLatency += tMs;
        latencyCount++;
      }
    }

    return {
      total_calls: totalCalls,
      total_cost: Math.round(totalCost * 1e6) / 1e6,
      avg_duration: Math.round((totalDuration / totalCalls) * 100) / 100,
      avg_latency_ms: latencyCount > 0
        ? Math.round((totalLatency / latencyCount) * 10) / 10
        : 0,
      cost_breakdown: {
        stt: Math.round(costStt * 1e6) / 1e6,
        tts: Math.round(costTts * 1e6) / 1e6,
        llm: Math.round(costLlm * 1e6) / 1e6,
        telephony: Math.round(costTel * 1e6) / 1e6,
      },
      active_calls: this.activeCalls.size,
    };
  }

  /** Return calls whose `started_at` falls within `[fromTs, toTs]` (Unix seconds). */
  getCallsInRange(fromTs = 0, toTs = 0): CallRecord[] {
    return this.calls.filter((call) => {
      const started = call.started_at || 0;
      if (fromTs && started < fromTs) return false;
      if (toTs && started > toTs) return false;
      return true;
    });
  }

  /** Number of completed calls currently in the ring buffer. */
  get callCount(): number {
    return this.calls.length;
  }

  /**
   * Rebuild the in-memory call list from `metadata.json` files written by
   * `CallLogger` under `<logRoot>/calls/YYYY/MM/DD/<call_id>/`. Idempotent:
   * call_ids already in the store are skipped. Errors per file are logged
   * and swallowed so a single corrupt entry doesn't block hydration.
   *
   * Returns the number of calls newly added to the store.
   *
   * Safe to call before any traffic; intended to run once at server startup.
   */
  hydrate(logRoot: string | null | undefined): number {
    if (!logRoot) return 0;
    const callsRoot = path.join(logRoot, 'calls');
    if (!fs.existsSync(callsRoot)) return 0;

    const collected: CallRecord[] = [];
    const seen = new Set<string>(this.calls.map((c) => c.call_id));

    const walk = (dir: string, depth: number): void => {
      let entries: fs.Dirent[];
      try {
        entries = fs.readdirSync(dir, { withFileTypes: true });
      } catch {
        return;
      }
      for (const entry of entries) {
        const childPath = path.join(dir, entry.name);
        if (depth < 3) {
          // YYYY / MM / DD layers — descend only when name is numeric to
          // skip stray files (DS_Store, indexes).
          if (entry.isDirectory() && /^\d+$/.test(entry.name)) {
            walk(childPath, depth + 1);
          }
          continue;
        }
        // depth === 3 → this is a per-call directory.
        if (!entry.isDirectory()) continue;
        const metadataPath = path.join(childPath, 'metadata.json');
        if (!fs.existsSync(metadataPath)) continue;
        try {
          const raw = fs.readFileSync(metadataPath, 'utf8');
          const meta = JSON.parse(raw) as Record<string, unknown>;
          const callId = (meta.call_id as string) || entry.name;
          if (!callId || seen.has(callId)) continue;
          const record = metadataToCallRecord(callId, meta);
          if (record === null) {
            // Unparseable started_at → skip rather than insert as epoch 0
            // (which would corrupt sort order forever).
            getLogger().debug(
              `MetricsStore.hydrate: skipping ${metadataPath}: unparseable started_at`,
            );
            continue;
          }
          collected.push(record);
          seen.add(callId);
        } catch (err) {
          getLogger().debug(
            `MetricsStore.hydrate: skipping ${metadataPath}: ${String(err)}`,
          );
        }
      }
    };

    walk(callsRoot, 0);

    // Stable order: oldest first (matches the order recordCallEnd would use).
    collected.sort((a, b) => (a.started_at || 0) - (b.started_at || 0));

    // Re-check seen against this.calls before each insert. Defends against the
    // (rare) case where hydrate() is invoked concurrently with itself or with
    // live recordCallEnd() traffic — without this guard, the snapshot taken
    // at the top of hydrate() can be stale by the time we reach the writes.
    for (const rec of collected) {
      if (this.calls.some((c) => c.call_id === rec.call_id)) continue;
      this.calls.push(rec);
      if (this.calls.length > this.maxCalls) {
        this.calls = this.calls.slice(-this.maxCalls);
      }
    }
    return collected.length;
  }
}

/**
 * Translate a CallLogger ``metadata.json`` payload into a ``CallRecord``.
 * Returns ``null`` when ``started_at`` is missing or unparseable — the record
 * would otherwise be silently inserted with ``started_at = 0`` (Unix epoch),
 * which corrupts every sort/range query that depends on it.
 */
function metadataToCallRecord(
  callId: string,
  meta: Record<string, unknown>,
): CallRecord | null {
  const startedAt = parseTimestamp(meta.started_at);
  if (startedAt === null) return null;
  const endedAt = parseTimestamp(meta.ended_at);
  const status = (meta.status as string | undefined) || 'completed';
  const metrics =
    meta.metrics && typeof meta.metrics === 'object'
      ? (meta.metrics as Record<string, unknown>)
      : null;
  const transcript = Array.isArray(meta.transcript)
    ? (meta.transcript as CallRecord['transcript'])
    : [];
  return {
    call_id: callId,
    caller: (meta.caller as string) || '',
    callee: (meta.callee as string) || '',
    direction: (meta.direction as string) || 'inbound',
    started_at: startedAt,
    ended_at: endedAt ?? undefined,
    status,
    metrics,
    transcript,
  };
}

/**
 * Parse a metadata timestamp into Unix seconds. Accepts numbers (seconds)
 * and ISO-8601 strings; returns ``null`` for missing, unrecognized, or
 * unparseable values so callers can decide to skip the record rather than
 * silently insert it as epoch 0.
 */
function parseTimestamp(raw: unknown): number | null {
  if (typeof raw === 'number') {
    return Number.isFinite(raw) ? raw : null;
  }
  if (typeof raw === 'string') {
    const ms = Date.parse(raw);
    return Number.isFinite(ms) ? ms / 1000 : null;
  }
  return null;
}

File: libraries/python/getpatter/dashboard/store.py
"""In-memory metrics store for the local dashboard."""

from __future__ import annotations

__all__ = ["MetricsStore", "MetricsStoreProtocol"]

import asyncio
import threading
import time
from dataclasses import asdict
from typing import Any, Protocol, runtime_checkable

from getpatter.models import CallMetrics


@runtime_checkable
class MetricsStoreProtocol(Protocol):
    """Protocol for metrics storage backends.

    B2B customers can implement this protocol to plug in their own
    metrics backend (e.g. Prometheus, Datadog, a database) while
    keeping the same dashboard interface.
    """

    def record_call_start(self, data: dict[str, Any]) -> None: ...

    def record_call_end(
        self, data: dict[str, Any], metrics: CallMetrics | None = None
    ) -> None: ...

    def record_turn(self, data: dict[str, Any]) -> None: ...

    def record_call_initiated(self, data: dict[str, Any]) -> None: ...

    def update_call_status(self, call_id: str, status: str, **extra: Any) -> None: ...

    def get_calls(self, limit: int = 50, offset: int = 0) -> list[dict[str, Any]]: ...

    def get_call(self, call_id: str) -> dict[str, Any] | None: ...

    def get_active_calls(self) -> list[dict[str, Any]]: ...

    def get_aggregates(self) -> dict[str, Any]: ...

    def subscribe(self) -> asyncio.Queue: ...

    def unsubscribe(self, queue: asyncio.Queue) -> None: ...


class MetricsStore:
    """Thread-safe in-memory store for call metrics.

    Keeps the last ``max_calls`` completed calls and tracks active calls.
    Designed as a singleton attached to the EmbeddedServer.

    Supports SSE event subscribers for real-time updates.
    """

    def __init__(self, max_calls: int = 500) -> None:
        self._lock = threading.Lock()
        self._max_calls = max_calls
        self._calls: list[dict[str, Any]] = []
        self._active_calls: dict[str, dict[str, Any]] = {}
        self._subscribers: set[asyncio.Queue] = set()

    # --- SSE event bus ---

    def subscribe(self) -> asyncio.Queue:
        """Subscribe to real-time events. Returns an asyncio.Queue."""
        queue: asyncio.Queue = asyncio.Queue(maxsize=100)
        with self._lock:
            self._subscribers.add(queue)
        return queue

    def unsubscribe(self, queue: asyncio.Queue) -> None:
        """Remove an SSE subscriber."""
        with self._lock:
            self._subscribers.discard(queue)

    def _publish(self, event_type: str, data: dict[str, Any]) -> None:
        """Push an event to all SSE subscribers (non-blocking).

        Must be called OUTSIDE self._lock to avoid deadlock with
        subscribe()/unsubscribe().
        """
        event = {"type": event_type, "data": data}
        # Snapshot to avoid iteration over a mutating set
        subscribers = list(self._subscribers)
        dead: list[asyncio.Queue] = []
        for q in subscribers:
            try:
                q.put_nowait(event)
            except asyncio.QueueFull:
                dead.append(q)
        if dead:
            with self._lock:
                for q in dead:
                    self._subscribers.discard(q)

    # --- Recording ---

    def record_call_start(self, data: dict[str, Any]) -> None:
        """Record the moment a call's media stream begins (publishes ``call_start``)."""
        call_id = data.get("call_id", "")
        if not call_id:
            return
        event_data = {
            "call_id": call_id,
            "caller": data.get("caller", ""),
            "callee": data.get("callee", ""),
            "direction": data.get("direction", "inbound"),
        }
        with self._lock:
            existing = self._active_calls.get(call_id)
            # If the call was pre-registered with ``record_call_initiated``
            # (e.g., outbound dial before media arrives), upgrade its status
            # to "in-progress" instead of overwriting the from/to metadata.
            # Only overwrite ``direction`` when the caller explicitly passed
            # one in ``data`` — otherwise we'd clobber the ``outbound`` set
            # by ``record_call_initiated`` with the default ``inbound``.
            if existing is not None:
                update_payload = {
                    "call_id": event_data["call_id"],
                    "caller": event_data["caller"],
                    "callee": event_data["callee"],
                }
                if "direction" in data:
                    update_payload["direction"] = data["direction"]
                existing.update(update_payload)
                existing["status"] = "in-progress"
                existing.setdefault("turns", [])
            else:
                self._active_calls[call_id] = {
                    **event_data,
                    "started_at": time.time(),
                    "status": "in-progress",
                    "turns": [],
                }
        # Publish outside lock to avoid deadlock with subscribe/unsubscribe
        self._publish("call_start", event_data)

    def record_call_initiated(self, data: dict[str, Any]) -> None:
        """Pre-register an outbound call before any webhook fires.

        Called from ``Patter.call()`` so that even calls that never reach the
        media channel (busy, no-answer, carrier-rejected) show up in the
        dashboard.
        """
        call_id = data.get("call_id", "")
        if not call_id:
            return
        entry = {
            "call_id": call_id,
            "caller": data.get("caller", ""),
            "callee": data.get("callee", ""),
            "direction": data.get("direction", "outbound"),
            "started_at": time.time(),
            "status": "initiated",
            "turns": [],
        }
        with self._lock:
            # Don't clobber a pre-existing record (e.g. inbound already moved
            # to "in-progress"). First writer wins.
            self._active_calls.setdefault(call_id, entry)
        self._publish(
            "call_initiated",
            {
                k: entry[k]
                for k in ("call_id", "caller", "callee", "direction", "status")
            },
        )

    def update_call_status(self, call_id: str, status: str, **extra: Any) -> None:
        """Update the status of an active or completed call.

        Used by the Twilio ``statusCallback`` handler and by operators
        wiring custom provider webhooks. Known statuses:
        ``initiated``, ``ringing``, ``in-progress``, ``completed``,
        ``no-answer``, ``busy``, ``failed``, ``canceled``, ``webhook_error``.
        """
        if not call_id or not status:
            return
        terminal = status in {
            "completed",
            "no-answer",
            "busy",
            "failed",
            "canceled",
            "webhook_error",
        }
        with self._lock:
            active = self._active_calls.get(call_id)
            if active is not None:
                active["status"] = status
                if extra:
                    active.update(extra)
                if terminal:
                    # Move to completed list so the UI stops the live timer.
                    entry = {
                        "call_id": call_id,
                        "caller": active.get("caller", ""),
                        "callee": active.get("callee", ""),
                        "direction": active.get("direction", "outbound"),
                        "started_at": active.get("started_at", 0),
                        "ended_at": time.time(),
                        "status": status,
                        "metrics": None,
                        **{k: v for k, v in extra.items() if k not in {"status"}},
                    }
                    self._active_calls.pop(call_id, None)
                    self._calls.append(entry)
                    if len(self._calls) > self._max_calls:
                        self._calls = self._calls[-self._max_calls :]
            else:
                # Call already completed — patch the existing row if found.
                for call in reversed(self._calls):
                    if call.get("call_id") == call_id:
                        call["status"] = status
                        if extra:
                            call.update(extra)
                        break
        self._publish("call_status", {"call_id": call_id, "status": status, **extra})

    def record_turn(self, data: dict[str, Any]) -> None:
        """Append a completed conversation turn to the active call (publishes ``turn_complete``)."""
        call_id = data.get("call_id", "")
        turn = data.get("turn")
        if not call_id or turn is None:
            return
        turn_dict = asdict(turn) if hasattr(turn, "__dataclass_fields__") else turn
        with self._lock:
            active = self._active_calls.get(call_id)
            if active is not None:
                active["turns"].append(turn_dict)
        # Publish outside lock to avoid deadlock with subscribe/unsubscribe
        self._publish("turn_complete", {"call_id": call_id, "turn": turn_dict})

    def record_call_end(
        self, data: dict[str, Any], metrics: CallMetrics | None = None
    ) -> None:
        """Move a call from the active set into history with final metrics (publishes ``call_end``)."""
        call_id = data.get("call_id", "")
        if not call_id:
            return
        with self._lock:
            active = self._active_calls.pop(call_id, None)
            entry: dict[str, Any] = {
                "call_id": call_id,
                "ended_at": time.time(),
                "transcript": data.get("transcript", []),
            }
            if active:
                entry["caller"] = active.get("caller", "")
                entry["callee"] = active.get("callee", "")
                entry["direction"] = active.get("direction", "inbound")
                entry["started_at"] = active.get("started_at", 0)
                # Preserve any explicit status (no-answer, busy, ...) set by
                # a statusCallback during the call. Fall back to "completed".
                entry["status"] = (
                    active.get("status", "completed")
                    if active.get("status") != "in-progress"
                    else "completed"
                )
            else:
                entry.setdefault("status", "completed")
            if metrics is not None:
                entry["metrics"] = asdict(metrics)
            else:
                # No metrics payload (e.g. webhook-rejected inbound, or
                # outbound call that never hit media): synthesise a minimal
                # metrics shim so the UI can still display a frozen duration.
                started = entry.get("started_at") or 0
                ended = entry.get("ended_at") or time.time()
                entry["metrics"] = {
                    "duration_seconds": max(0.0, float(ended - started)),
                    "turns": [],
                    "cost": {
                        "total": 0.0,
                        "stt": 0.0,
                        "tts": 0.0,
                        "llm": 0.0,
                        "telephony": 0.0,
                    },
                    "latency_avg": {"total_ms": 0.0},
                    "latency_p50": {"total_ms": 0.0},
                    "latency_p90": {"total_ms": 0.0},
                    "latency_p95": {"total_ms": 0.0},
                    "latency_p99": {"total_ms": 0.0},
                    "provider_mode": "",
                }
            self._calls.append(entry)
            if len(self._calls) > self._max_calls:
                self._calls = self._calls[-self._max_calls :]
            event_metrics = entry.get("metrics")
        # Publish outside lock to avoid deadlock with subscribe/unsubscribe
        self._publish(
            "call_end",
            {
                "call_id": call_id,
                "metrics": event_metrics,
            },
        )

    def get_calls(self, limit: int = 50, offset: int = 0) -> list[dict[str, Any]]:
        """Return the most recent completed calls, newest first."""
        with self._lock:
            ordered = list(reversed(self._calls))
            return ordered[offset : offset + limit]

    def get_call(self, call_id: str) -> dict[str, Any] | None:
        """Return the completed-call record for ``call_id`` if present."""
        with self._lock:
            for call in reversed(self._calls):
                if call["call_id"] == call_id:
                    return call
            return None

    def get_active_calls(self) -> list[dict[str, Any]]:
        """Return the currently in-flight calls."""
        with self._lock:
            return list(self._active_calls.values())

    def get_active(self, call_id: str) -> dict[str, Any] | None:
        """Return the active-call record for ``call_id`` if present.

        Mirrors the TypeScript ``MetricsStore.getActive`` accessor used by
        handlers that need to peek at the live call (e.g., to read the
        current ``direction`` without taking a write lock).
        """
        with self._lock:
            return self._active_calls.get(call_id)

    def get_aggregates(self) -> dict[str, Any]:
        """Compute aggregate stats (call count, cost, avg duration, latency) across history."""
        with self._lock:
            total_calls = len(self._calls)
            if total_calls == 0:
                return {
                    "total_calls": 0,
                    "total_cost": 0.0,
                    "avg_duration": 0.0,
                    "avg_latency_ms": 0.0,
                    "cost_breakdown": {
                        "stt": 0.0,
                        "tts": 0.0,
                        "llm": 0.0,
                        "telephony": 0.0,
                    },
                    "active_calls": len(self._active_calls),
                }

            total_cost = 0.0
            total_duration = 0.0
            total_latency = 0.0
            latency_count = 0
            cost_stt = 0.0
            cost_tts = 0.0
            cost_llm = 0.0
            cost_tel = 0.0

            for call in self._calls:
                m = call.get("metrics")
                if m is None:
                    continue
                cost = m.get("cost", {})
                total_cost += cost.get("total", 0.0)
                cost_stt += cost.get("stt", 0.0)
                cost_tts += cost.get("tts", 0.0)
                cost_llm += cost.get("llm", 0.0)
                cost_tel += cost.get("telephony", 0.0)
                total_duration += m.get("duration_seconds", 0.0)
                avg_lat = m.get("latency_avg", {})
                t_ms = avg_lat.get("total_ms", 0.0)
                if t_ms > 0:
                    total_latency += t_ms
                    latency_count += 1

            return {
                "total_calls": total_calls,
                "total_cost": round(total_cost, 6),
                "avg_duration": round(total_duration / total_calls, 2),
                "avg_latency_ms": round(total_latency / latency_count, 1)
                if latency_count > 0
                else 0.0,
                "cost_breakdown": {
                    "stt": round(cost_stt, 6),
                    "tts": round(cost_tts, 6),
                    "llm": round(cost_llm, 6),
                    "telephony": round(cost_tel, 6),
                },
                "active_calls": len(self._active_calls),
            }

    def get_calls_in_range(
        self, from_ts: float = 0.0, to_ts: float = 0.0
    ) -> list[dict[str, Any]]:
        """Return calls within a timestamp range (inclusive)."""
        with self._lock:
            result = []
            for call in self._calls:
                started = call.get("started_at", 0)
                if from_ts and started < from_ts:
                    continue
                if to_ts and started > to_ts:
                    continue
                result.append(call)
            return result

    @property
    def call_count(self) -> int:
        """Number of completed calls currently held in memory."""
        with self._lock:
            return len(self._calls)

    def hydrate(self, log_root: str | None) -> int:
        """Rebuild the call list from on-disk metadata.json files.

        ``CallLogger`` persists per-call envelopes under
        ``<log_root>/calls/YYYY/MM/DD/<call_id>/metadata.json``. Calling
        ``hydrate(log_root)`` once at server startup replays those files
        into the in-memory store so the dashboard survives a restart (the
        durable persistence is the JSONL/JSON files; this store is just a
        cache on top).

        Idempotent: ``call_id``s already in the store are skipped. Errors
        per file are logged at debug level and swallowed so a single
        corrupt entry doesn't block hydration.

        Returns the number of calls newly added.
        """
        import json
        import logging
        from pathlib import Path

        if not log_root:
            return 0
        calls_root = Path(log_root) / "calls"
        if not calls_root.is_dir():
            return 0

        log = logging.getLogger("getpatter.dashboard.store")
        collected: list[dict[str, Any]] = []
        with self._lock:
            seen = {c.get("call_id") for c in self._calls if c.get("call_id")}

        for year in _numeric_subdirs(calls_root):
            for month in _numeric_subdirs(year):
                for day in _numeric_subdirs(month):
                    for call_dir in day.iterdir():
                        if not call_dir.is_dir():
                            continue
                        meta_path = call_dir / "metadata.json"
                        if not meta_path.is_file():
                            continue
                        try:
                            with open(meta_path, encoding="utf-8") as fh:
                                meta = json.load(fh)
                        except (OSError, json.JSONDecodeError) as exc:
                            log.debug(
                                "MetricsStore.hydrate: skipping %s: %s",
                                meta_path,
                                exc,
                            )
                            continue
                        call_id = meta.get("call_id") or call_dir.name
                        if not call_id or call_id in seen:
                            continue
                        record = _metadata_to_call_record(call_id, meta)
                        if record is None:
                            # Unparseable started_at → skip rather than insert
                            # as epoch 0 (which would corrupt sort order).
                            log.debug(
                                "MetricsStore.hydrate: skipping %s: "
                                "unparseable started_at",
                                meta_path,
                            )
                            continue
                        collected.append(record)
                        seen.add(call_id)

        # Stable order: oldest first (matches recordCallEnd's append order).
        collected.sort(key=lambda r: r.get("started_at") or 0)

        # Re-check call_id presence under the lock before each insert.
        # Defends against the rare case where hydrate() is invoked
        # concurrently with itself or with live recordCallEnd traffic.
        with self._lock:
            existing_ids = {c.get("call_id") for c in self._calls if c.get("call_id")}
            for rec in collected:
                if rec["call_id"] in existing_ids:
                    continue
                self._calls.append(rec)
                existing_ids.add(rec["call_id"])
                if len(self._calls) > self._max_calls:
                    self._calls = self._calls[-self._max_calls :]
        return len(collected)


def _numeric_subdirs(parent):
    """Yield direct subdirectories of ``parent`` whose name is all digits."""
    try:
        entries = list(parent.iterdir())
    except OSError:
        return
    for entry in entries:
        if entry.is_dir() and entry.name.isdigit():
            yield entry


def _metadata_to_call_record(
    call_id: str, meta: dict[str, Any]
) -> dict[str, Any] | None:
    """Translate a CallLogger metadata.json payload into a CallRecord dict.

    Returns ``None`` when ``started_at`` is missing or unparseable — the
    record would otherwise be silently inserted with ``started_at = 0``
    (Unix epoch), which corrupts every sort/range query that depends on it.
    """
    from datetime import datetime

    def _to_seconds(raw: Any) -> float | None:
        if isinstance(raw, (int, float)):
            return float(raw)
        if isinstance(raw, str):
            try:
                return datetime.fromisoformat(raw.replace("Z", "+00:00")).timestamp()
            except ValueError:
                return None
        return None

    started = _to_seconds(meta.get("started_at"))
    if started is None:
        return None
    ended = _to_seconds(meta.get("ended_at"))
    metrics = meta.get("metrics") if isinstance(meta.get("metrics"), dict) else None
    transcript = (
        meta.get("transcript") if isinstance(meta.get("transcript"), list) else []
    )
    record: dict[str, Any] = {
        "call_id": call_id,
        "caller": meta.get("caller") or "",
        "callee": meta.get("callee") or "",
        "direction": meta.get("direction") or "inbound",
        "started_at": started,
        "status": meta.get("status") or "completed",
        "metrics": metrics,
        "transcript": transcript,
    }
    if ended is not None:
        record["ended_at"] = ended
    return record
